### PR TITLE
Fix CGR featurizer behavior when bond features matrix is empty

### PR DESCRIPTION
**Pull request by jonwzheng**
## Description
Previously Chemprop would crash while featurizing certain reactions with the following error:
```
ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 53 has 1 dimension(s)
```
@KnathanM found that this is due to a bug with `CondensedGraphOfReactionFeaturizer`, which, for molecular graphs with no edges, instantiates `E` as an empty list, which cannot be concatenated to the other feature matrices, leading to the crash.

## Bugfix / Desired workflow
This PR changes the featurizer code so that `E` will be instantiated with the correct number of dimensions if it is empty.

This code also gets rid of an unused iteration variable.

## Checklist
- [x] linted with flake8?

### CODE REVIEW

File: chemprop/featurizers/molgraph/reaction.py [hunk 1]
```diff
@@ -135,9 +134,7 @@ def __call__(
                 edge_index[0].extend([u, v])
                 edge_index[1].extend([v, u])
 
-                i += 2
-
-        E = np.array(E)
+        E = np.array(E) if E else np.empty((0, self.bond_fdim))
```
Comment (davidegraff) [inline chemprop/featurizers/molgraph/reaction.py:137]: implicit calls to `bool()` are best avoided unless you _know_ it's the behavior you want.

Comment (jonwzheng) [reply chemprop/featurizers/molgraph/reaction.py:137]: good point - how about `E = np.array(E) if len(E) > 0 else ` ...  ?

**Review by shihchengli — CHANGES_REQUESTED** [review]
Can you add a test case in the unit test (tests/unit/featurizers/test_cgr.py)?

File: chemprop/featurizers/molgraph/reaction.py (hunk 1 above)
Comment (davidegraff) [reply chemprop/featurizers/molgraph/reaction.py]: I'd have to assume that's the exact behavior you want to be testing, so makes sense to me

**Comment by jonwzheng** [comment]
Great idea! I added a unit test as you suggested. 
I'm not very happy with it though - it was kind of tricky because of how the test suite is set up for CGR. 
There are two "hacky" things I welcome suggestions for improving, I'll comment directly on the code pointing them out.

File: tests/unit/featurizers/test_cgr.py [hunk 2]
```diff
@@ -105,6 +107,9 @@ def rxn_smi(request):
         BondExpectation((0, 2), bond_reac_none=True, bond_prod_none=True),
         BondExpectation((1, 2), bond_reac_none=False, bond_prod_none=False),
     ],
+    "[CH4:1]>>[CH3:1].[H:2]": [
+        BondExpectation((0, 0), bond_reac_none=True, bond_prod_none=True),
```
Comment (jonwzheng) [inline tests/unit/featurizers/test_cgr.py:111]: this felt a little hacky to me because it's not really a bond. And I don't know if this is expected to remain the default behavior forever.

**Review by davidegraff — COMMENTED** [review]
a minor suggestion here is to change `BondExpectation` to a class using the `typing.NamedTuple` class-based syntax

File: tests/unit/featurizers/test_cgr.py [hunk 3]
```diff
@@ -381,5 +392,13 @@ def test_featurize_imbalanced(self, rxn_smi, mode_imbalanced):
         for i, bond in enumerate(bonds):
             expect_edge_index.extend([bond, bond[::-1]])
             expect_rev_edge_index.extend([i * 2 + 1, i * 2])
-        assert np.array_equal(molgraph.edge_index, expect_edge_index)
-        assert np.array_equal(molgraph.rev_edge_index, expect_rev_edge_index)
+
+        if len(expect_edge_index) == 0:
```
Comment (jonwzheng) [inline tests/unit/featurizers/test_cgr.py:396]: this also felt unsatisfying - but this is basically because `molgraph.edge_index` is instantiated as an np array of `[ [] [] ]` (a list of at least two lists) whereas `expect_edge_index` is constructed by extending an empty list (`[]`). So in the null case, the molgraph edge will be an np array with 2 dimensions, but in the latter it will just be 1 dimensional, so they fail the `np._array_equal` test.

I welcome your feedback on how to write this more elegantly/cleanly if necessary!

File: tests/unit/featurizers/test_cgr.py (hunk 2 above)
Comment (shihchengli) [reply tests/unit/featurizers/test_cgr.py]: This is the way RDKit is implemented. I think it is fine. Especially, our code would never use u and v as the same number. Will this test reach the code you changed in reaction.py? If not, I suggest we remove this test here.

File: tests/unit/featurizers/test_cgr.py (hunk 3 above)
Comment (shihchengli) [reply tests/unit/featurizers/test_cgr.py:396]: Maybe initiate it the same way as in `molgraph.edge_index`

**Review by shihchengli — COMMENTED** [review]
I found bugs in `test_get_bonds_imbalanced` and `test_get_bonds_balanced`. The assertions lack proper indentation. Can you also adjust it?

File: tests/unit/featurizers/test_cgr.py (hunk 2 above)
Comment (jonwzheng) [reply tests/unit/featurizers/test_cgr.py]: I put this here because of the way the test suite is set up: right now, anything in `rxn_smis` is expected to also have an example in `reac_prod_maps` and in `bond_expect_imbalanced`. If we remove this test, we will have to add some exceptions which might make the testing suite unnecessarily more complicated. But, I agree this doesn't really test anything meaningful. Let me know what you think.

File: tests/unit/featurizers/test_cgr.py (hunk 3 above)
Comment (jonwzheng) [reply tests/unit/featurizers/test_cgr.py:395]: thanks! I changed it now.

File: tests/unit/featurizers/test_cgr.py (hunk 2 above)
Comment (shihchengli) [reply tests/unit/featurizers/test_cgr.py]: Without making much change, it seems this is the only way we can do this test. Could you add a comment on this line to mention this is only for test purposes for graphs with zero edges?

**Comment by jonwzheng** [comment]
nice catch @shihchengli, thanks for sharing that. I addressed your + david's comments in the latest commits.

File: tests/unit/featurizers/test_cgr.py (hunk 2 above)
Comment (jonwzheng) [reply tests/unit/featurizers/test_cgr.py]: done!

**Review by shihchengli — APPROVED** [review]

**Comment by jonwzheng** [comment]
@davidegraff just checking in to see if these changes look good to you since you left a review comment previously

**Review by davidegraff — APPROVED** [review]
